runtime/near-vm-runner/src/runner.rs

```diff
@@ -146,3 +149,77 @@ impl VMKindExt for VMKind {
         }
     }
 }
+
+pub struct FilesystemCompiledContractCache {
+    dir: rustix::fd::OwnedFd,
+}
+
+impl FilesystemCompiledContractCache {
+    pub fn new() -> Self {
+        let path = std::env::var_os("NEAR_CONTRACT_CACHE");
+        let path = path.map(PathBuf::from).unwrap_or_else(|| {
+            PathBuf::from(std::env::var_os("HOME").expect("$HOME"))
+                .join(".near")
+                .join("data")
+                .join("contracts")
+        });
+        std::fs::create_dir_all(&path).expect("create contract directory");
+        let dir = rustix::fs::open(path, rustix::fs::OFlags::DIRECTORY, rustix::fs::Mode::empty())
+            .expect("open contract dir");
+        Self { dir }
+    }
+}
+
+/// Cache for compiled contracts code in plain filesystem.
+impl CompiledContractCache for FilesystemCompiledContractCache {
+    fn put(&self, key: &CryptoHash, value: CompiledContract) -> std::io::Result<()> {
+        use rustix::fs::{Mode, OFlags};
+        let final_filename = key.to_string();
+        let filename = format!("{}.temp", key);
+        let mode = Mode::RUSR | Mode::WUSR | Mode::RGRP | Mode::WGRP;
+        let flags = OFlags::CREATE | OFlags::TRUNC | OFlags::WRONLY;
+        // FIXME: we probably want to make a file with a temporary name here first (use mktemp
+        // crate) and then move it into the expected place.
+        // FIXME: remove the file once done if it still exists.
+        let mut file = std::fs::File::from(rustix::fs::openat(&self.dir, &filename, flags, mode)?);
+        match value {
+            CompiledContract::CompileModuleError(_) => {
+                // TODO: ideally we probably want this to go into a separate file, or use some sort
```

Comment:
Maybe the comments should clarify that we are intentionally leaving the file empty in this case.  A reader might suspect that there is some missing code here.